# Chapter 2 The Mathmatical Building blocks of neural network

## Chapter 2.1 First Look at a neural work

Before that let's check if the tensorflow has been installed correctly


First of all , check if the GPU version has been installed

In [2]:
import tensorflow as tf 
tf.test.gpu_device_name()

'/device:GPU:0'

In [3]:
print("The tensorflow version is {}".format(tf.__version__))

The tensorflow version is 2.9.1


In [4]:
tf.__path__

['e:\\Python\\lib\\site-packages\\keras\\api\\_v2',
 'e:\\Python\\lib\\site-packages\\tensorflow_estimator\\python\\estimator\\api\\_v2',
 'e:\\Python\\lib\\site-packages\\tensorboard\\summary\\_tf',
 'e:\\Python\\lib\\site-packages\\tensorflow',
 'e:\\Python\\lib\\site-packages\\tensorflow\\_api\\v2']

## 2.1 First Look at a neural network

### 2.2.1 Loading the MINST dataset in Keras

In [7]:
from tensorflow.keras.datasets import mnist 


In [9]:
(train_image,train_labels),(test_image,test_labels) = mnist.load_data()

In [13]:
print(train_image.shape)
print(test_image.shape)

(60000, 28, 28)
(10000, 28, 28)


In [14]:
print(train_labels.shape)
print(test_labels.shape)

(60000,)
(10000,)


### 2.1.2 The network architechure

In [15]:
from tensorflow import keras
from tensorflow.keras import layers


In [18]:
model = keras.Sequential([
    layers.Dense(512,activation='relu'),
    layers.Dense(10,activation='softmax')
])

The Core building block of nerual network is the layer

Layers extract representations out of the data fed into them 

Here our model consists of a sequence of two Dense layers, which are densely connected (fully connected) neural layers

The second is a 10-way softmax classification layer, which means it will return an array of 10 probabilty scores (sum=1)
Each score will be the probabilty that the current digit image belongs to one of our 10 digit classes

To make model ready for training. We need to pick 3 more things as a part of compilation step
1. An optimizer: The mechanism through which the model will update itself based on the training data
2. A loss function: How model will be able to measure its performance on the training data
3. Metrics: To monitor during training and testing. Here we use accuracy. 

### 2.1.3 The compilation step

In [20]:
model.compile(
            optimizer = 'rmsprop',
            loss = 'sparse_categorical_crossentropy',
            metrics =['accuracy']
            )

### 2.1.4 Preparing the image data

In [21]:
train_images = train_image.reshape((60000,28*28))
train_images = train_images.astype('float32')/255

test_images =test_image.reshape((10000,28*28))
test_images =test_images.astype('float32')/255


### 2.1.5 Fitting the model

In [22]:
model.fit(train_images,train_labels,epochs=5,batch_size=128)


Epoch 1/5
469/469 [==============================] - 6s 7ms/step - loss: 0.2529 - accuracy: 0.9272
Epoch 2/5
469/469 [==============================] - 3s 7ms/step - loss: 0.1029 - accuracy: 0.9696
Epoch 3/5
469/469 [==============================] - 3s 7ms/step - loss: 0.0674 - accuracy: 0.9798
Epoch 4/5
469/469 [==============================] - 3s 7ms/step - loss: 0.0487 - accuracy: 0.9852
Epoch 5/5
469/469 [==============================] - 3s 7ms/step - loss: 0.0366 - accuracy: 0.9887


We get cost = 0.0366 and accuracy = 0.9887 here

### 2.1.6 Using the model to make predictions

In [23]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
predictions[0]

1/1 [==============================] - 0s 91ms/step


array([3.1791498e-09, 4.3488602e-11, 1.7360996e-07, 1.4839533e-06,
       2.2513779e-12, 9.3203623e-10, 4.5336825e-14, 9.9999833e-01,
       2.9862143e-10, 3.5015120e-08], dtype=float32)

Each number of index i in the arrat correspondes to the probabilty that digit image test_digits[0] belongs to class

In [24]:
predictions[0].argmax()


7

In [25]:
predictions[0][7]

0.99999833

We can see that the maximum probablity belongs to class 7, which means this digit should be 7

We can check the  first test label

In [26]:
test_labels[0] 

7

In [27]:
test_labels[0] == predictions[0].argmax()

True

Proves that what we have predicted is correct !

In [32]:
prediction_all = model.predict(test_images)
prediction_all = [pred.argmax() for pred in prediction_all]


313/313 [==============================] - 1s 2ms/step


We can check the score on test set

In [33]:
import numpy as np 

np.mean((prediction_all == test_labels))

0.9788

### 2.1.7 Evaluating the model on new data

In [34]:
test_loss,test_acc = model.evaluate(test_images,test_labels)
print("Test Accuracy {:.2f}".format(test_acc))

313/313 [==============================] - 2s 5ms/step - loss: 0.0690 - accuracy: 0.9788
Test Accuracy 0.98


## 2.2 Data Represents for neural network